docs/en/practice_todos_app/develop_a_todos_app.md

```diff
@@ -0,0 +1,694 @@
+
+```rust ignore
+// todos.rs
+
+use serde::{Deserialize, Serialize};
+use std::{
+  collections::BTreeMap,
+  fs::File,
+  io::{self, BufWriter, Write},
+};
+
+#[derive(Debug, Serialize, Deserialize)]
+pub struct Todos {
+  tasks: BTreeMap<TaskId, Task>,
+  next_id: TaskId,
+}
+
+#[derive(Debug, Clone, PartialEq, Serialize, Deserialize)]
+pub struct Task {
+  id: TaskId,
+  pub complete: bool,
+  pub label: String,
+}
+
+#[derive(Debug, Clone, Copy, PartialEq, Eq, PartialOrd, Ord, Serialize, Deserialize)]
+pub struct TaskId(usize);
+
+impl Todos {
+  pub fn new_task(&mut self, label: String) -> TaskId {
+    let id = self.next_id;
+    self.next_id = self.next_id.next();
+    self.tasks.insert(id, Task { id, label, complete: false });
+    
+    id
+  }
+
+  pub fn remove(&mut self, id: TaskId) { self.tasks.remove(&id); }
+
+  pub fn get_task(&self, id: TaskId) -> Option<&Task> { self.tasks.get(&id) }
+
+  pub fn get_task_mut(&mut self, id: TaskId) -> Option<&mut Task> { self.tasks.get_mut(&id) }
+
+  pub fn all_tasks(&self) -> impl Iterator<Item = TaskId> + '_ { self.tasks.keys().copied() }
+}
+
+impl Task {
+  pub fn id(&self) -> TaskId { self.id }
+}
+
+impl Todos {
+  pub fn load() -> Self {
+    std::fs::read(Self::store_path())
+      .map(|v| serde_json::from_slice(v.as_slice()).unwrap())
+      .unwrap_or_else(|_| Todos {
+        tasks: BTreeMap::new(),
+        next_id: TaskId(0),
+      })
+  }
+
+  pub fn save(&self) -> Result<(), io::Error> {
+    let file = File::create(Self::store_path())?;
+    let mut writer = BufWriter::new(file);
+    serde_json::to_writer(&mut writer, self)?;
+    writer.flush()?;
+    Ok(())
+  }
+
+  fn store_path() -> std::path::PathBuf { std::env::temp_dir().join("ribir_todos.json") }
+}
+
+impl TaskId {
+  pub fn next(&self) -> Self { Self(self.0 + 1) }
+}
+```
+
+The `Todos` core is mainly composed of three types: `Todos`, `Task`, and `TaskId`. `Todos` is a collection containing all tasks, `Task` is a struct for a task, and `TaskId` is a unique identifier for a task. `Todos` provides methods for adding, deleting, modifying, and querying tasks, as well as the ability to save to a file. Usually, you also need to write comprehensive unit tests to ensure the correctness of your code.
+
+This part of the work is no different from how you usually write Rust code without a GUI. You can organize your code according to your own habits, as long as you can provide a complete API in the end. In the design philosophy of Ribir applications, this part of the work is very important, but it is not the focus of this tutorial. If you are familiar with Rust syntax, you should be able to understand it easily, so we won't go into detail here.
```

Comment:
In the design philosophy of Ribir applications, this part of the work is very important, but it is not the focus of this tutorial. If you are familiar with Rust syntax, you should be able to understand it easily, so we won't go into detail here. 
-> In the design philosophy of Ribir applications, this part of the work is very important, but it is not the focus of this tutorial, so we won't go into detail here.